### PR TITLE
Update mypy job to 1.1.1

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/pull/16933#discussion_r1124397243

cc @carmocca @akihironitta @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/core/datamodule.py [hunk 1]
```diff
@@ -223,11 +223,12 @@ def load_from_checkpoint(
...
-        return _load_from_checkpoint(
+        loaded = _load_from_checkpoint(
             cls,
             checkpoint_path,
             map_location=None,
             hparams_file=hparams_file,
             strict=None,
             **kwargs,
         )
+        return cast(Self, loaded)
```
Comment (carmocca) [inline src/lightning/pytorch/core/datamodule.py:234]: I tried to avoid this cast, but it would really complicate the `_load_from_checkpoint` typing. We would need TypeVars bounded to LM and LDM + maybe some overloads. Couldn't get it to work.

Another solution would be to inline the `_load_from_checkpoint` implementation here (and for the LDM), so that the return type is not ambiguous

File: src/lightning/pytorch/core/module.py [hunk 2]
```diff
@@ -1419,6 +1439,96 @@ def to_torchscript(
 
         return torchscript_module
 
+    @classmethod
+    def load_from_checkpoint(
```
Comment (carmocca) [inline src/lightning/pytorch/core/module.py:1443]: ModelIO is removed, all the code was put here. The only class that uses it

Comment (awaelchli) [reply src/lightning/pytorch/core/module.py:1443]: Since it was public api, needs to be documented in changelog

File: src/lightning/pytorch/serve/servable_module.py [hunk 3]
```diff
@@ -56,9 +57,11 @@ def configure_response(self):
         assert serve_cb.resp.json() == {"output": [0, 1]}
     """
 
+    @abstractmethod
```
Comment (awaelchli) [inline src/lightning/pytorch/serve/servable_module.py:60]: shouldn't this be done in a different pr? Seems unrelated to typing

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/accelerators/ipu.py [hunk 4]
```diff
@@ -49,9 +49,9 @@ def parse_devices(devices: int) -> int:
         return devices
 
     @staticmethod
-    def get_parallel_devices(devices: int) -> List[int]:
+    def get_parallel_devices(devices: int) -> List[torch.device]:
```
Comment (carmocca) [inline src/lightning/pytorch/accelerators/ipu.py:52]: I believe all accelerators should return a list of torch.devices, which is what the strategy expects to get as `strategy.parallel_devices`

Otherwise there can be silent errors where the wrong device type is used.

File: src/lightning/pytorch/serve/servable_module.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/pytorch/serve/servable_module.py:60]: Mypy was complaining that these methods annotate that they return a Dict when the true return is None (empty body)

File: src/lightning/pytorch/core/module.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/pytorch/core/module.py:1443]: Good call. Done

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 5]
```diff
@@ -179,7 +179,7 @@ def is_custom_reduction(self) -> bool:
         return not (self.is_mean_reduction or self.is_max_reduction or self.is_min_reduction or self.is_sum_reduction)
 
 
-class _ResultMetric(Metric, _DeviceDtypeModuleMixin):
+class _ResultMetric(Metric, _DeviceDtypeModuleMixin):  # type: ignore[misc]  # torchmetrics methods should return Self
```
Comment (Borda) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:182]: cc: @SkafteNicki @justusschock

**Review by Borda — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]
